packages/fixie-sdk/src/request-wrapper.tsx

```diff
@@ -13,6 +13,22 @@ export const RequestContext = AI.createContext<{
   agentId: string;
 } | null>(null);
 
+/**
+ * Renders to "in-progress" while the children are still being rendered, and "done" when they're done.
+ */
+async function* MessageState({ children }: { children: AI.Node }, { render }: AI.ComponentContext) {
+  const renderResult = render(children);
+  let didYield = false;
+  for await (const _ of renderResult) {
+    if (!didYield) {
+      didYield = true;
+      yield 'in-progress';
+    }
+  }
+
+  return 'done';
```

Comment:
I can't 100% verify that this works correctly due my limited knowledge of `for await` details. So the idea is that up until `renderResult` is yielding we stay in the `for` loop and once it actually `returns` the loop is done?
If you've verified this I'd take your word for it.

Reply:
Another good question. The short answer is that yes, that's how the `render` API works, but your question prompted me to wonder whether this might effectively introduce [screen tearing](https://en.wikipedia.org/wiki/Screen_tearing) because there's no guarantee that this will resolve _after_ the other render of the element resolves.

Need to think about that more :)

Reply:
Ok, I added the (slightly heavy-handed, but I think net "good") `batchFrames` option.

The main catch is that it can(/will) delay frames in favor of CPU-bound rendering work, but I suspect on the balance it will perform better in the cases we care about.